File: bharat-fm/sdks/javascript/src/domains/governance.ts
/**
 * Governance AI domain-specific client for BFMF
 */

import { BharatAIClient } from '../client';
import { GenerationRequest, GenerationResponse } from '../types';
import { BharatAIError } from '../errors';

/**
 * RTI (Right to Information) request
 */
export interface RTIRequest {
  /** RTI application text */
  applicationText: string;
  /** Department/Ministry */
  department: string;
  /** Request type */
  requestType: 'information' | 'inspection' | 'sample';
  /** Urgency level */
  urgency?: 'normal' | 'urgent' | 'immediate';
  /** Language */
  language?: string;
}

/**
 * RTI response
 */
export interface RTIResponse {
  /** Generated RTI response */
  responseText: string;
  /** Original application */
  originalApplication: string;
  /** Department */
  department: string;
  /** Response time estimate */
  responseTimeEstimate: string;
  /** Relevant sections */
  relevantSections: Array<{
    section: string;
    description: string;
    relevance: 'high' | 'medium' | 'low';
  }>;
  /** Confidence score */
  confidence: number;
  /** Processing time */
  processingTime: number;
}

/**
 * Policy analysis request
 */
export interface PolicyAnalysisRequest {
  /** Policy document text */
  policyText: string;
  /** Analysis type */
  analysisType: 'summary' | 'impact' | 'compliance' | 'stakeholder' | 'implementation';
  /** Sector/Domain */
  sector?: string;
  /** Target audience */
  targetAudience?: string;
  /** Language */
  language?: string;
}

/**
 * Policy analysis response
 */
export interface PolicyAnalysisResponse {
  /** Analysis results */
  analysis: string;
  /** Key insights */
  keyInsights: string[];
  /** Recommendations */
  recommendations: string[];
  /** Risk assessment */
  riskAssessment?: {
    level: 'low' | 'medium' | 'high';
    factors: string[];
  };
  /** Compliance score */
  complianceScore?: number;
  /** Analysis time */
  analysisTime: number;
}

/**
 * Compliance audit request
 */
export interface ComplianceAuditRequest {
  /** Document to audit */
  documentText: string;
  /** Compliance framework */
  framework: string;
  /** Audit type */
  auditType: 'full' | 'targeted' | 'risk-based';
  /** Specific requirements to check */
  requirements?: string[];
  /** Industry/sector */
  industry?: string;
}

/**
 * Compliance audit response
 */
export interface ComplianceAuditResponse {
  /** Overall compliance score */
  complianceScore: number;
  /** Detailed findings */
  findings: Array<{
    requirement: string;
    status: 'compliant' | 'non-compliant' | 'partial';
    evidence: string;
    severity: 'low' | 'medium' | 'high' | 'critical';
    recommendation: string;
  }>;
  /** Summary */
  summary: string;
  /** Action items */
  actionItems: string[];
  /** Audit time */
  auditTime: number;
}

/**
 * Government scheme analysis request
 */
export interface SchemeAnalysisRequest {
  /** Scheme name or description */
  scheme: string;
  /** Analysis type */
  analysisType: 'eligibility' | 'benefits' | 'application' | 'comparison';
  /** User context */
  userContext?: {
    income?: string;
    category?: string;
    location?: string;
    occupation?: string;
  };
  /** Language */
  language?: string;
}

/**
 * Government scheme analysis response
 */
export interface SchemeAnalysisResponse {
  /** Scheme details */
  schemeDetails: {
    name: string;
    description: string;
    ministry: string;
    category: string;
  };
  /** Analysis results */
  analysis: string;
  /** Eligibility criteria */
  eligibilityCriteria?: string[];
  /** Benefits */
  benefits?: string[];
  /** Application process */
  applicationProcess?: string[];
  /** Similar schemes */
  similarSchemes?: Array<{
    name: string;
    similarity: number;
    keyDifferences: string[];
  }>;
  /** Analysis time */
  analysisTime: number;
}

/**
 * Governance AI domain-specific client
 */
export class GovernanceAIClient {
  private client: BharatAIClient;
  private domainEndpoint: string;

  /**
   * Create a new Governance AI client
   */
  constructor(client: BharatAIClient, domainEndpoint: string = '/governance') {
    this.client = client;
    this.domainEndpoint = domainEndpoint;
  }

  /**
   * Generate RTI response
   */
  async generateRTIResponse(request: RTIRequest): Promise<RTIResponse> {
    const response = await this.client['makeRequest']<RTIResponse>({
      method: 'POST',
      url: `${this.domainEndpoint}/rti-response`,
      data: {
        application_text: request.applicationText,
        department: request.department,
        request_type: request.requestType,
        urgency: request.urgency || 'normal',
        language: request.language
      }
    });

    return response;
  }

  /**
   * Analyze policy document
   */
  async analyzePolicy(request: PolicyAnalysisRequest): Promise<PolicyAnalysisResponse> {
    const response = await this.client['makeRequest']<PolicyAnalysisResponse>({
      method: 'POST',
      url: `${this.domainEndpoint}/analyze-policy`,
      data: {
        policy_text: request.policyText,
        analysis_type: request.analysisType,
        sector: request.sector,
        target_audience: request.targetAudience,
        language: request.language
      }
    });

    return response;
  }

  /**
   * Conduct compliance audit
   */
  async conductComplianceAudit(request: ComplianceAuditRequest): Promise<ComplianceAuditResponse> {
    const response = await this.client['makeRequest']<ComplianceAuditResponse>({
      method: 'POST',
      url: `${this.domainEndpoint}/compliance-audit`,
      data: {
        document_text: request.documentText,
        framework: request.framework,
        audit_type: request.auditType,
        requirements: request.requirements,
        industry: request.industry
      }
    });

    return response;
  }

  /**
   * Analyze government scheme
   */
  async analyzeScheme(request: SchemeAnalysisRequest): Promise<SchemeAnalysisResponse> {
    const response = await this.client['makeRequest']<SchemeAnalysisResponse>({
      method: 'POST',
      url: `${this.domainEndpoint}/analyze-scheme`,
      data: {
        scheme: request.scheme,
        analysis_type: request.analysisType,
        user_context: request.userContext,
        language: request.language
      }
    });

    return response;
  }

  /**
   * Batch RTI response generation
   */
  async generateRTIResponseBatch(requests: RTIRequest[]): Promise<RTIResponse[]> {
    const response = await this.client['makeRequest']<RTIResponse[]>({
      method: 'POST',
      url: `${this.domainEndpoint}/rti-response-batch`,
      data: requests.map(req => ({
        application_text: req.applicationText,
        department: req.department,
        request_type: req.requestType,
        urgency: req.urgency || 'normal',
        language: req.language
      }))
    });

    return response;
  }

  /**
   * Get list of supported departments
   */
  async getSupportedDepartments(): Promise<Array<{
    name: string;
    code: string;
    description: string;
    categories: string[];
  }>> {
    const response = await this.client['makeRequest']<Array<{
      name: string;
      code: string;
      description: string;
      categories: string[];
    }>>({
      method: 'GET',
      url: `${this.domainEndpoint}/departments`
    });

    return response;
  }

  /**
   * Get compliance frameworks
   */
  async getComplianceFrameworks(): Promise<Array<{
    name: string;
    description: string;
    industry: string;
    version: string;
    requirements: string[];
  }>> {
    const response = await this.client['makeRequest']<Array<{
      name: string;
      description: string;
      industry: string;
      version: string;
      requirements: string[];
    }>>({
      method: 'GET',
      url: `${this.domainEndpoint}/compliance-frameworks`
    });

    return response;
  }

  /**
   * Get government schemes database
   */
  async getGovernmentSchemes(filters?: {
    ministry?: string;
    category?: string;
    beneficiary?: string;
    state?: string;
  }): Promise<Array<{
    name: string;
    description: string;
    ministry: string;
    category: string;
    beneficiary: string;
    eligibility: string[];
    benefits: string[];
    applicationProcess: string[];
    website?: string;
    contact?: string;
  }>> {
    const response = await this.client['makeRequest']<Array<{
      name: string;
      description: string;
      ministry: string;
      category: string;
      beneficiary: string;
      eligibility: string[];
      benefits: string[];
      applicationProcess: string[];
      website?: string;
      contact?: string;
    }>>({
      method: 'GET',
      url: `${this.domainEndpoint}/schemes`,
      params: filters
    });

    return response;
  }

  /**
   * Generate policy draft
   */
  async generatePolicyDraft(params: {
    title: string;
    scope: string;
    objectives: string[];
    stakeholders: string[];
    existingPolicies?: string[];
    language?: string;
  }): Promise<{
    draft: string;
    sections: Array<{
      title: string;
      content: string;
      importance: 'high' | 'medium' | 'low';
    }>;
    estimatedReviewTime: string;
  }> {
    const response = await this.client['makeRequest']<{
      draft: string;
      sections: Array<{
        title: string;
        content: string;
        importance: 'high' | 'medium' | 'low';
      }>;
      estimatedReviewTime: string;
    }>({
      method: 'POST',
      url: `${this.domainEndpoint}/generate-policy-draft`,
      data: params
    });

    return response;
  }

  /**
   * Track RTI application status
   */
  async trackRTIStatus(rtiId: string): Promise<{
    status: 'received' | 'processing' | 'responded' | 'rejected' | 'appealed';
    currentDepartment: string;
    estimatedResponseDate?: string;
    lastUpdate: string;
    updates: Array<{
      date: string;
      status: string;
      remarks: string;
      officer?: string;
    }>;
  }> {
    const response = await this.client['makeRequest']<{
      status: 'received' | 'processing' | 'responded' | 'rejected' | 'appealed';
      currentDepartment: string;
      estimatedResponseDate?: string;
      lastUpdate: string;
      updates: Array<{
        date: string;
        status: string;
        remarks: string;
        officer?: string;
      }>;
    }>({
      method: 'GET',
      url: `${this.domainEndpoint}/rti-status/${rtiId}`
    });

    return response;
  }
}

File: bharat-fm/sdks/go/internal/domains/governance.go
package domains

import (
    "context"
    "fmt"

    "github.com/bharat-ai/bharat-fm/sdks/go/internal/client"
    "github.com/bharat-ai/bharat-fm/sdks/go/internal/models"
)

// GovernanceAIClient provides governance-specific AI capabilities
type GovernanceAIClient struct {
    client *client.Client
}

// NewGovernanceAIClient creates a new governance AI client
func NewGovernanceAIClient(client *client.Client) *GovernanceAIClient {
    return &GovernanceAIClient{
        client: client,
    }
}

// RTIRequest represents an RTI request
type RTIRequest struct {
    ApplicationText string `json:"application_text"`
    Department      string `json:"department"`
    RequestType     string `json:"request_type"`
    Urgency         string `json:"urgency,omitempty"`
    Language        string `json:"language,omitempty"`
}

// RTIResponse represents an RTI response
type RTIResponse struct {
    ResponseText        string `json:"response_text"`
    OriginalApplication string `json:"original_application"`
    Department          string `json:"department"`
    ResponseTimeEstimate string `json:"response_time_estimate"`
    RelevantSections    []struct {
        Section     string `json:"section"`
        Description string `json:"description"`
        Relevance   string `json:"relevance"`
    } `json:"relevant_sections"`
    Confidence float64 `json:"confidence"`
    ProcessingTime float64 `json:"processing_time"`
}

// PolicyAnalysisRequest represents a policy analysis request
type PolicyAnalysisRequest struct {
    PolicyText     string `json:"policy_text"`
    AnalysisType   string `json:"analysis_type"`
    Sector         string `json:"sector,omitempty"`
    TargetAudience string `json:"target_audience,omitempty"`
    Language       string `json:"language,omitempty"`
}

// PolicyAnalysisResponse represents a policy analysis response
type PolicyAnalysisResponse struct {
    Analysis         string   `json:"analysis"`
    KeyInsights     []string `json:"key_insights"`
    Recommendations  []string `json:"recommendations"`
    RiskAssessment  *struct {
        Level   string   `json:"level"`
        Factors []string `json:"factors"`
    } `json:"risk_assessment,omitempty"`
    ComplianceScore float64 `json:"compliance_score,omitempty"`
    AnalysisTime    float64 `json:"analysis_time"`
}

// GenerateRTIResponse generates an RTI response
func (c *GovernanceAIClient) GenerateRTIResponse(ctx context.Context, req *RTIRequest) (*RTIResponse, error) {
    var result RTIResponse
    
    apiReq := map[string]interface{}{
        "application_text": req.ApplicationText,
        "department":       req.Department,
        "request_type":     req.RequestType,
    }
    
    if req.Urgency != "" {
        apiReq["urgency"] = req.Urgency
    }
    if req.Language != "" {
        apiReq["language"] = req.Language
    }

    resp, err := c.client.GetHttpClient().R().
        SetContext(ctx).
        SetBody(apiReq).
        SetResult(&result).
        Post("/governance/rti-response")
    
    if err != nil {
        return nil, fmt.Errorf("RTI response request failed: %w", err)
    }
    
    if resp.StatusCode() != 200 {
        return nil, fmt.Errorf("RTI response API error: status %d", resp.StatusCode())
    }
    
    return &result, nil
}

// AnalyzePolicy analyzes a policy document
func (c *GovernanceAIClient) AnalyzePolicy(ctx context.Context, req *PolicyAnalysisRequest) (*PolicyAnalysisResponse, error) {
    var result PolicyAnalysisResponse
    
    apiReq := map[string]interface{}{
        "policy_text":   req.PolicyText,
        "analysis_type": req.AnalysisType,
    }
    
    if req.Sector != "" {
        apiReq["sector"] = req.Sector
    }
    if req.TargetAudience != "" {
        apiReq["target_audience"] = req.TargetAudience
    }
    if req.Language != "" {
        apiReq["language"] = req.Language
    }

    resp, err := c.client.GetHttpClient().R().
        SetContext(ctx).
        SetBody(apiReq).
        SetResult(&result).
        Post("/governance/analyze-policy")
    
    if err != nil {
        return nil, fmt.Errorf("policy analysis request failed: %w", err)
    }
    
    if resp.StatusCode() != 200 {
        return nil, fmt.Errorf("policy analysis API error: status %d", resp.StatusCode())
    }
    
    return &result, nil
}

// GetSupportedDepartments gets supported departments
func (c *GovernanceAIClient) GetSupportedDepartments(ctx context.Context) ([]struct {
    Name        string   `json:"name"`
    Code        string   `json:"code"`
    Description string   `json:"description"`
    Categories  []string `json:"categories"`
}, error) {
    var result []struct {
        Name        string   `json:"name"`
        Code        string   `json:"code"`
        Description string   `json:"description"`
        Categories  []string `json:"categories"`
    }

    resp, err := c.client.GetHttpClient().R().
        SetContext(ctx).
        SetResult(&result).
        Get("/governance/departments")
    
    if err != nil {
        return nil, fmt.Errorf("get departments request failed: %w", err)
    }
    
    if resp.StatusCode() != 200 {
        return nil, fmt.Errorf("get departments API error: status %d", resp.StatusCode())
    }
    
    return result, nil
}

// SimpleGenerateRTIResponse provides a simple RTI response interface
func (c *GovernanceAIClient) SimpleGenerateRTIResponse(ctx context.Context, applicationText, department string) (string, error) {
    req := &RTIRequest{
        ApplicationText: applicationText,
        Department:      department,
        RequestType:     "information",
    }
    
    resp, err := c.GenerateRTIResponse(ctx, req)
    if err != nil {
        return "", err
    }
    
    return resp.ResponseText, nil
}

// SimpleAnalyzePolicy provides a simple policy analysis interface
func (c *GovernanceAIClient) SimpleAnalyzePolicy(ctx context.Context, policyText, analysisType string) (string, error) {
    req := &PolicyAnalysisRequest{
        PolicyText:   policyText,
        AnalysisType: analysisType,
    }
    
    resp, err := c.AnalyzePolicy(ctx, req)
    if err != nil {
        return "", err
    }
    
    return resp.Analysis, nil
}